components/blitz/src/omero/cmd/fs/UsedFilesRequestI.java

```diff
@@ -0,0 +1,369 @@
+        @SuppressWarnings("unchecked")
+        final List<Object> results = query.list();
+        if (results.isEmpty()) {
+            final Exception e = new IllegalArgumentException("cannot read image " + imageId);
+            throw helper.cancel(new ERR(), e, "bad-image");
+        }
+        if (results.get(0) != null) {
+            filesetId = (Long) results.get(0);
+        } else if (isFromPlate()) {
+            final Exception e = new IllegalArgumentException("image " + imageId + " is from a pre-FS plate");
+            throw helper.cancel(new ERR(), e, "bad-image");
+        }
+        try {
+            findPixels();
+        } catch (Cancel c) {
+            throw c;
+        } catch (Throwable t) {
+            throw helper.cancel(new ERR(), t, "used-files-fail");
+        }
+        return filesetId == null ? determineResponsePreFs() : determineResponseFs();
+    }
+
+    /**
+     * Find the image's primary pixels object and, if an FS image, open a Bio-Formats reader for it.
+     * @throws Cancel if there are no pixels for the image
+     * @throws FormatException if the pixels service could not open a Bio-Formats reader for the image
+     * @throws IOException if the pixels service could not open a Bio-Formats reader for the image
+     */
+    private void findPixels() throws Cancel, FormatException, IOException {
+        final String hql = "FROM Image WHERE id = :id";
+        final Query query = session.createQuery(hql).setParameter("id", imageId);
+        final Image image = (Image) query.uniqueResult();
+        if (image.sizeOfPixels() < 1) {
+            final Exception e = new IllegalArgumentException("no pixels for image " + imageId);
+            throw helper.cancel(new ERR(), e, "bad-image");
+        }
+        final Pixels pixels = image.getPrimaryPixels();
+        pixelsId = pixels.getId();
+        if (filesetId != null) {
+            reader = pixelsService.getBfReader(pixels);
+        }
+    }
+
+    /**
+     * @param repo a repository identifier
+     * @return the root of the repository on the filesystem
+     */
+    private File getRepositoryRoot(String repo) {
+        File root = repositoryRoots.get(repo);
+        if (root == null) {
+            final String hql = "SELECT path || name FROM OriginalFile WHERE mimetype = 'Repository' AND hash = :repo";
+            final Query query = session.createQuery(hql).setParameter("repo", repo);
+            final String repositoryRootDirectory = (String) query.uniqueResult();
+            root = new File(repositoryRootDirectory);
+            repositoryRoots.put(repo, root);
+        }
+        return root;
+    }
+
+    /**
+     * @param repo a repository
+     * @param file a file path in the repository
+     * @return the path of the file on the filesystem
+     */
+    private File getUnderlyingFile(String repo, String file) {
+        /* note: cannot use ServerFilePathTransformer from within a Request */
```

Comment:
True, they could be moved, though right now they are happily sitting in what feels like an appropriate place. Luckily in this case doing without them was rather less mess than I feared because the paths I find in the database have already been through all that machinery once on their way there.